lib/srv/db/redis/connection.go

```diff
@@ -0,0 +1,147 @@
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// DefaultPort is the Redis default port.
+const DefaultPort = "6379"
+
+const (
+	// URIScheme is a Redis scheme: https://www.iana.org/assignments/uri-schemes/prov/redis
+	// Teleport always uses Redis connection over TLS.
+	URIScheme = "redis"
+	// URISchemeTLS is a Redis scheme that uses TLS for database connection: https://www.iana.org/assignments/uri-schemes/prov/rediss
+	URISchemeTLS = "rediss"
+)
+
+// ConnectionMode defines the mode in which Redis is configured. Currently, supported are single and cluster.
+type ConnectionMode string
+
+const (
+	// Standalone mode should be used when connecting to a single Redis instance.
+	Standalone ConnectionMode = "standalone"
+	// Cluster mode should be used then connecting to a Redis Cluster.
+	Cluster ConnectionMode = "cluster"
+)
+
+// ConnectionOptions defines Redis connection options.
+type ConnectionOptions struct {
+	// mode defines Redis connection mode like cluster or single instance.
+	mode ConnectionMode
+	// address of Redis instance.
+	address string
+	// port on which Redis expects new connections.
+	port string
+}
+
+// ParseRedisAddress parses a Redis connection string and returns the parsed
+// connection options like address and connection mode. If port is skipped
+// default Redis 6379 is used.
+// Correct inputs:
+// 	rediss://redis.example.com:6379?mode=cluster
+// 	redis://redis.example.com:6379
+// 	redis.example.com:6379
+//
+// Incorrect input:
+//	redis.example.com:6379?mode=cluster
+func ParseRedisAddress(addr string) (*ConnectionOptions, error) {
+	if addr == "" {
+		return nil, trace.BadParameter("Redis address is empty")
+	}
+
+	var redisURL url.URL
+	var instanceAddr string
+
+	if strings.Contains(addr, "://") {
+		// Assume URI version
+		u, err := url.Parse(addr)
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse Redis URI: %q, error: %v", addr, err)
+		}
+
+		switch u.Scheme {
+		case URIScheme, URISchemeTLS:
+		default:
+			return nil, trace.BadParameter("failed to parse Redis address %q, invalid Redis URI scheme: %q. "+
+				"Expected %q or %q.", addr, u.Scheme, URIScheme, URISchemeTLS)
+		}
+
+		redisURL = *u
+		instanceAddr = u.Host
+	} else {
+		// Assume host:port pair
+		instanceAddr = addr
+	}
+
+	var (
+		host string
+		port string
+	)
+
+	if strings.Contains(instanceAddr, ":") {
+		var err error
+		host, port, err = net.SplitHostPort(instanceAddr)
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse Redis host: %q, error: %v", addr, err)
+		}
+
+		// Check if the port can be parsed as a number. net.SplitHostPort() doesn't guarantee that.
+		_, err = strconv.Atoi(port)
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse Redis URL %q, please provide instance address in "+
+				"form address:port or rediss://address:port, error: %v", addr, err)
+		}
+
+	} else {
+		port = DefaultPort
+		host = instanceAddr
+	}
```

Comment:
nit: i feel this parsing is a common use case. maybe we can beef up our `lib/utils/addr.go` to have a version that mandates port must be an integer (or maybe update one of the existing). That may help others in the future who needs the same check.

Reply:
I didn't want to do that as a part of this PR as that change may affect more components than just Redis integration (you would expect that this check won't break anything, but you never know 🤷‍♂️ )
I can "beefipy" our current implementation as a separate PR and ask people from other teams to review it. Does it sound ok?

Reply:
no worries. I think we can update utils when we actually need it.